Add role option for record_desktop_session

zmb3 opened the pull request:
This option defaults to true, as defined in RFD #0033.
The preset editor and audit roles disable desktop session recording,
as they aren't permitted access to desktops. The preset access role
enables desktop session recording.

zmb3 committed:
Add role option for record_desktop_session

This option defaults to true, as defined in RFD #0033.

The preset editor and audit roles disable desktop session recording,
as they aren't permitted access to desktops. The preset access role
enables desktop session recording. The implicit role applied to all
role sets also disables recording, otherwise it would be impossible
to disable.

zmb3 commented on lib/services/role.go:
```diff
@@ -1691,6 +1691,17 @@ func (set RoleSet) CanPortForward() bool {
 
+// RecordDesktopSession returns true if a role in the role set has enabled
+// desktop session recoring.
+func (set RoleSet) RecordDesktopSession() bool {
```
Note to reviewers: this uses the same logic that we use throughout the RBAC system for  boolean options that default to true. That is, as long as one of the user's roles has enabled the feature, it will be enabled.

I *think* this is what we want here, but it does mean it may be difficult to disable session recording, as *all* of the user's roles would have to explicitly disable it.

Seem reasonable or might we need a more flexible approach here?

ibeckermayer replied:
Good question. I was just looking through the code and documentation and it looks like we don't truly have precedence for this feature with SSH session recording, since it's set at the cluster-level and so there aren't such role-level conflicts.

This raises the question -- do we truly want role-level configurability for session recording? I can see it being a useful feature, but barring a specific request from a customer I think we should mimic SSH session recording and configure it in `windows_desktop_service` to start.

zmb3 replied:
Yes, we truly want it. This was discussed and approved in RFD 33.

In a nutshell, the reason for having this associated with roles as opposed to cluster wide is that some users may have long (several hours) desktop sessions and not want to save the very large recording (not an issue with smaller SSH recordings).

rosstimothy replied:
Will there be accompanying docs changes that highlight this difference in behavior? I could see users being slightly confused about this at first.

ibeckermayer replied:
@rosstimothy FTR we're currently discussing this in an email thread with Sasha and Xin. Slow going right now due to Sasha being OOO.

ibeckermayer reviewed (commented):
We may want to take a different approach, see my inline comment

https://github.com/gravitational/teleport/pull/9523#discussion_r775555672

ibeckermayer reviewed (approved):
For the tests: 
- Consider making the `"full valid role"` test have `"record_desktop_session": false` in order to confirm that `record_desktop_session` is parsed properly.
- Consider adding a `RoleSet` test that demonstrates that any `RoleSet` where a single `record_desktop_session` is `true` (and/or unset) results in the setting being set to `true`.

zmb3 commented:
Note: not planning to backport this, as session recording will be a Teleport 9 feature and not part of the Desktop Access preview in Teleport 8.

klizhentas reviewed (commented):
@zmb3 is there any reason why we need to differentiate protocol in `record_desktop_session`?  How would that scale for other protocols if we were to support the switch?

What do you think of this format?

```yaml
record_session:
    desktop: true
    ssh: false
```

zmb3 commented:
@klizhentas I don't have a strong preference, just going with what was approved in the RFD.

SSH session recording is configured cluster-wide in the `SessionRecordingConfig` resource. I don't think changing that is in scope here, but is something we can look at in a follow up change.

If we think that's a change we're likely to make, then your proposed format makes sense. Do we have evidence of people wanting to be able to configure SSH recordings on a per-role basis?

zmb3 committed:
Apply Sasha's suggestion for a record_session property

This allows to eventually record SSH sessions based on role options
in the same way as desktop sessions.

zmb3 committed:
Merge branch 'master' into zmb3/desktop-session-role-updates
